text/3735-field-projections.md

```diff
@@ -0,0 +1,1789 @@
+- Feature Name: `field_projection`
+- Start Date: 2024-10-24
+- RFC PR: [rust-lang/rfcs#3735](https://github.com/rust-lang/rfcs/pull/3735)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Field projections are a very general concept. In simple terms, it is a new operator that turns a
```

Comment:
This is a bit of a nit: in the context of this RFC the field projection operation is only applicable to reference-like containers. For example consider the case of `Arc<T>`: suppose I have some code that wants an `Arc<F>`, and then I refactor things such that I now have an `Arc<T>` instead, where `T` has a field of type `F`; this proposal does not cover the case of owned value projection - I cannot obtain an `Arc<F>` from my `Arc<T>`. Right?

There's some prior art for this kind of projection: https://docs.rs/yoke/latest/yoke/.

Reply:
> This is a bit of a nit: in the context of this RFC the field projection operation is only applicable to reference-like containers. For example consider the case of `Arc<T>`: suppose I have some code that wants an `Arc<F>`, and then I refactor things such that I now have an `Arc<T>` instead, where `T` has a field of type `F`; this proposal does not cover the case of owned value projection - I cannot obtain an `Arc<F>` from my `Arc<T>`. Right?

Nothing prevents us from implementing `Project<F> for Arc<T>`, but I think it'd be a footgun, since you have to construct an entirely new `Arc<F::Type>`. Since an `Arc<T>` is essentially just a `NonNull<ArcInner<T>>`, we can just look at the layout of that:
```rust
#[repr(C)]
struct Foo {
    a: usize,
    b: usize,
}

// layout of `Foo`:
//
// |a-------|b-------|

// layout of `ArcInner<Foo>`
//
// |strong--|weak----|data.a--|data.b--|

// layout of `ArcInner<usize>`
//
// |strong--|weak----|value---|
```

The problem is now that we *have* to create a new `ArcInner<usize>` on the heap, copy the value from `Foo` and then drop the old `Arc`. There are two reasons:
- if the field is not the first one, we have to move it, since the `ArcInner` expects the value *directly* after the weak refcount
- in order for `Arc` to reliably free the memory it has allocated, it needs to know the size of the data. If we were to just transmute `NonNull<ArcInner<Foo>>` to `NonNull<ArcInner<usize>>` when projecting to the field `a`, then the `free` function of the allocator could be called with the wrong layout.


Overall, I feel like this is a rather complex operation (and I haven't really used it in code). So it shouldn't be hidden away behind an operator, instead a method taking a generic parameter that is a field type would be a much better fit in my opinion.


Do you think that the RFC is unclear w.r.t. this? If yes, then I'll gladly add the above to it.

> There's some prior art for this kind of projection: https://docs.rs/yoke/latest/yoke/.

Oh yeah, I forgot about that crate! But I think that it also can only provide projections for the yoke type. So if you have `Yoke<Bar<'static>, Rc<[u8]>>` then you can project to its fields (considering the [example from the docs](https://docs.rs/yoke/latest/yoke/struct.Yoke.html#method.map_project)), but the cart type `Rc<[u8]>` will be unaffected.

Reply:
Thanks for the thorough answer! It all makes sense to me. It might be good to include a section explaining why such transmutation is not possible in the general case. Otherwise the phrasing implies too much generality.

Reply:
Oh one thing that I forgot to say was that if we change how `Arc` is implemented, then the projection you suggested above is possible (so if we have two pointers, one to the refcounters and another to the value). The RFC already talks about this in the [`ArcRef`](https://github.com/y86-dev/rfcs/blob/field-projection-v2/text/3735-field-projections.md#arcreft-for-stdlib) section.

But I would argue that `Arc<T>`/`ArcRef<T>` is a "reference-like container" type (taking your wording here, I would rather use the term "pointer-like type"), since it essentially is just a `NonNull<ArcInner<T>>`.

Reply:
Yep, makes sense - I don't think folks will be overly keen on changing how Arc is implemented since it increases the allocation cost.

`Arc<T>` is a "reference-like container" but its reference is to an `ArcInner<T>` which is not reference-like because it holds both the count and the allocation in-line. My suggested wording is probably not enough to clarify the point, but I still think the wording would benefit from some clarification, specifically that intrusive containers (e.g. `ArcInner<T>` - again please use better language if you have it) aren't appropriate for field projection.